driver-core/src/main/com/mongodb/connection/InternalStreamConnection.java

```diff
@@ -588,53 +522,28 @@ private MongoException translateReadException(final Throwable e) {
     }
 
     private ResponseBuffers receiveResponseBuffers() throws IOException {
-        ByteBuf messageHeaderBytBuf = stream.read(MessageHeader.MESSAGE_HEADER_LENGTH);
-        ByteBufferBsonInput headerInputBuffer = new ByteBufferBsonInput(messageHeaderBytBuf);
+        ByteBuf messageHeaderBuffer = stream.read(MESSAGE_HEADER_LENGTH);
         MessageHeader messageHeader;
         try {
-            messageHeader = new MessageHeader(headerInputBuffer, description.getMaxMessageSize());
+            messageHeader = new MessageHeader(messageHeaderBuffer, description.getMaxMessageSize());
         } finally {
-            headerInputBuffer.close();
+            messageHeaderBuffer.release();
         }
 
+        ByteBuf messageBuffer = stream.read(messageHeader.getMessageLength() - MESSAGE_HEADER_LENGTH);
```

Comment:
Now we just read the rest of the message, which includes the rest of the header and the body.  It was simpler that way.  Same on the async path.